src/repository/Repository.ts

```diff
@@ -1,21 +1,86 @@
+import { promisify } from "util";
 
 class Repository {
-    public async getAllUsers() {
-        return await Users.find();
+    private redisClient: RedisClient;
+    private getAsync: (key: string) => string;
+    private setAsync: (key: string, val: string) => void;
+    private getKeysAsync: (keys: string) => string[];
+
+    constructor() {
+        const redisOpts: ClientOpts = {
+            host: 'host',
+            port: 11316,
+            password: 'passwd'
+        };
+        this.redisClient = createClient(redisOpts);
+        this.getAsync = promisify(this.redisClient.get).bind(this.redisClient);
+        this.setAsync = promisify(this.redisClient.set).bind(this.redisClient);
+        this.getKeysAsync = promisify(this.redisClient.keys).bind(this.redisClient);
     }
 
-    public async saveRewardEvent(rewarder: User, rewardees: User[], channelID: string) {
-        const eventInstance = new RewardEvent();
-        eventInstance.rewarder = rewarder.id;
-        eventInstance.timestamp = new Date();
-        eventInstance.karma = 1;
-        eventInstance.channel = channelID;
+    public saveRewardEvent(rewarder: User, rewardees: User[], channelID: string) {
         rewardees.forEach(async (rewardee) => {
+            const eventInstance = new RewardEvent();
+            eventInstance.rewarder = rewarder.id;
+            eventInstance.timestamp = new Date();
+            eventInstance.karma = 1;
+            eventInstance.channel = channelID;
             eventInstance.rewardee = rewardee.id;
             await RewardEvent.save(eventInstance);
+            this.setRewardeeChannelKarma(rewardee.id, channelID);
+        });
+        
+    }
+
+    public async getLeaderboardInfo() {
+        const keys = await this.getKeysAsync('karmaInfo:*');
+        const karmaInfo: KarmaInfo[] = await Promise.all(keys.map(async (key) => {
+            const karma = await this.getAsync(key);
+            const [_, userID, channelID] = key.split(':');
+            return {
+                userID,
+                channelID,
+                karma: parseInt(karma)
+            }
+        }));
+        const userIDList = karmaInfo.map(info => info.userID);
+        
+        const guild = await client.guilds.fetch('842338736124854293');
+        const guildChannels = guild.channels.valueOf();
+        const userList = await guild.members.fetch({user: userIDList});
+        const userKarmaList = karmaInfo.map(info => {
+            const userName = userList.get(info.userID).user.username;
+            return {
+                user: userName,
+                karma: info.karma,
+                channel: guildChannels.get(info.channelID)?.name
+            }
+        });
+        return userKarmaList.sort((a, b) => b.karma - a.karma);
+    }
+
+    private setRewardeeChannelKarma(userID: string, channelID: string): void {
+        this.redisClient.GET(`karmaInfo:${userID}:${channelID}`, (err: Error, reply: string) => {
```

Comment:
we were planning to have a setup like `"karmaInfo:userId": { channelId1: karma, channelId2: karma }`. Any blockers with that approach?

Reply:
No blockers with the approach, but I don't think that will improve the performance or anything. Let's go with the current approach, we will improve it if the need be.

Reply:
cool ok